File: nodejs/ocr-client.js
/**
 * Meditrail OCR API Client - Node.js Example
 * 
 * This module demonstrates how to use the Meditrail OCR API to process
 * medical documents and images for clinical data extraction.
 */

const fs = require('fs');
const path = require('path');
const FormData = require('form-data');
const axios = require('axios');

/**
 * Client for Meditrail OCR API
 */
class MeditrailOCRClient {
    /**
     * Initialize the OCR client
     * 
     * @param {string} apiKey - Your Meditrail API key
     * @param {string} baseUrl - Base URL for the API (default: production URL)
     */
    constructor(apiKey, baseUrl = 'https://meditrail.unitedwecare.com/api/v1') {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash
        this.client = axios.create({
            timeout: 60000, // 60 second timeout
            headers: {
                'X-API-Key': apiKey
            }
        });
    }

    /**
     * Process a document using the OCR API
     * 
     * @param {string} filePath - Path to the file to process (PDF, JPG, PNG)
     * @param {string} [text] - Optional context or annotation
     * @param {string} [systemPrompt] - Optional prompt to customize extraction logic
     * @returns {Promise<Object>} API response
     * @throws {Error} If file operations or API request fails
     */
    async processDocument(filePath, text = null, systemPrompt = null) {
        // Validate file exists
        if (!fs.existsSync(filePath)) {
            throw new Error(`File not found: ${filePath}`);
        }

        // Check file size (50MB limit)
        const stats = fs.statSync(filePath);
        const fileSizeInMB = stats.size / (1024 * 1024);
        if (stats.size > 50 * 1024 * 1024) { // 50MB in bytes
            throw new Error(`File too large: ${fileSizeInMB.toFixed(2)}MB (max: 50MB)`);
        }

        // Prepare the request
        const url = `${this.baseUrl}/ocr/process`;
        
        // Create form data
        const formData = new FormData();
        formData.append('file', fs.createReadStream(filePath), {
            filename: path.basename(filePath)
        });

        if (text) {
            formData.append('text', text);
        }

        if (systemPrompt) {
            formData.append('system_prompt', systemPrompt);
        }

        try {
            const response = await this.client.post(url, formData, {
                headers: {
                    ...formData.getHeaders(),
                    'X-API-Key': this.apiKey
                }
            });

            return response.data;

        } catch (error) {
            this.handleError(error);
        }
    }

    /**
     * Handle API errors and throw appropriate exceptions
     * 
     * @param {Error} error - The error from axios
     * @throws {Error} With appropriate error message
     */
    handleError(error) {
        if (error.response) {
            // Server responded with error status
            const { status, data } = error.response;
            
            switch (status) {
                case 400:
                    const badRequestDetail = data.detail || 'Bad Request';
                    throw new Error(`Bad Request: ${badRequestDetail}`);
                case 413:
                    throw new Error('File too large (max 50MB)');
                case 429:
                    if (data.detail && typeof data.detail === 'object' && data.detail.message) {
                        throw new Error(`Usage limit exceeded: ${data.detail.message}`);
                    }
                    throw new Error('Usage limit exceeded');
                case 500:
                    const serverErrorDetail = data.detail || 'Internal Server Error';
                    throw new Error(`Internal Server Error: ${serverErrorDetail}`);
                default:
                    throw new Error(`Unexpected response: ${status} - ${JSON.stringify(data)}`);
            }
        } else if (error.request) {
            // Request was made but no response received
            throw new Error('No response received from server');
        } else {
            // Something else happened
            throw new Error(`Request setup error: ${error.message}`);
        }
    }
}

/**
 * Example usage of the Meditrail OCR Client
 */
async function main() {
    // Configuration
    const API_KEY = 'mt_LaRysGk4Ap9ytP29VKtF9E_8MECKcLcZSgRYLAF2KP5qr_UVPdCUqchtJ9fM7joiXX8z';
    
    // Initialize client
    const client = new MeditrailOCRClient(API_KEY);
    
    // Example 1: Process a medical image
    console.log('=== Example 1: Processing Medical Image ===');
    try {
        const filePath = 'sample_files/chest_xray.jpg';
        
        if (fs.existsSync(filePath)) {
            const result = await client.processDocument(
                filePath,
                'Chest X-ray examination',
                'Extract key clinical findings and abnormalities'
            );
            
            console.log(`Success! Document ID: ${result.id}`);
            console.log(`Clinical Relevance: ${result.clinical_relevance}`);
            console.log(`Doctor Names: ${result.doctor_names}`);
            
            // Parse the response field (it's a JSON string)
            const responseData = JSON.parse(result.response || '{}');
            console.log(`Document Type: ${responseData.document_type || 'N/A'}`);
            console.log(`Summary: ${responseData.summary || 'N/A'}`);
            
            // Print metadata
            const metadata = result.metadata || {};
            console.log(`Original File: ${metadata.original_file_name}`);
            console.log(`File Size: ${metadata.file_size}`);
            console.log(`Page Count: ${metadata.page_count}`);
            
        } else {
            console.log(`Sample file not found: ${filePath}`);
            console.log('Please add a sample medical document to test with.');
        }
        
    } catch (error) {
        console.error(`Error processing document: ${error.message}`);
    }
    
    console.log('\n' + '='.repeat(50) + '\n');
    
    // Example 2: Process a PDF document
    console.log('=== Example 2: Processing PDF Document ===');
    try {
        const filePath = 'sample_files/prescription.pdf';
        
        if (fs.existsSync(filePath)) {
            const result = await client.processDocument(
                filePath,
                'Prescription document',
                'Extract medication names, dosages, and doctor information'
            );
            
            console.log(`Success! Document ID: ${result.id}`);
            console.log(`Clinical Relevance: ${result.clinical_relevance}`);
            
            // Parse the response field
            const responseData = JSON.parse(result.response || '{}');
            console.log(`Document Type: ${responseData.document_type || 'N/A'}`);
            console.log(`Summary: ${responseData.summary || 'N/A'}`);
            
        } else {
            console.log(`Sample file not found: ${filePath}`);
            console.log('Please add a sample PDF document to test with.');
        }
        
    } catch (error) {
        console.error(`Error processing document: ${error.message}`);
    }
}

// Export the class for use in other modules
module.exports = MeditrailOCRClient;

// Run the example if this file is executed directly
if (require.main === module) {
    main().catch(console.error);
}

File: python/ocr_client.py
#!/usr/bin/env python3
"""
Meditrail OCR API Client - Python Example

This script demonstrates how to use the Meditrail OCR API to process
medical documents and images for clinical data extraction.
"""

import requests
import json
import os
import sys
from typing import Optional, Dict, Any
from pathlib import Path


class MeditrailOCRClient:
    """Client for Meditrail OCR API"""
    
    def __init__(self, api_key: str, base_url: str = "https://meditrail.unitedwecare.com/api/v1"):
        """
        Initialize the OCR client
        
        Args:
            api_key: Your Meditrail API key
            base_url: Base URL for the API (default: production URL)
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.headers = {
            'X-API-Key': api_key
        }
    
    def process_document(
        self, 
        file_path: str, 
        text: Optional[str] = None, 
        system_prompt: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Process a document using the OCR API
        
        Args:
            file_path: Path to the file to process (PDF, JPG, PNG)
            text: Optional context or annotation
            system_prompt: Optional prompt to customize extraction logic
            
        Returns:
            API response as dictionary
            
        Raises:
            FileNotFoundError: If the file doesn't exist
            requests.RequestException: If the API request fails
        """
        # Validate file exists
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"File not found: {file_path}")
        
        # Check file size (50MB limit)
        file_size = os.path.getsize(file_path)
        if file_size > 50 * 1024 * 1024:  # 50MB in bytes
            raise ValueError(f"File too large: {file_size / (1024*1024):.2f}MB (max: 50MB)")
        
        # Prepare the request
        url = f"{self.base_url}/ocr/process"
        
        # Prepare multipart form data
        files = {
            'file': (os.path.basename(file_path), open(file_path, 'rb'))
        }
        
        data = {}
        if text:
            data['text'] = text
        if system_prompt:
            data['system_prompt'] = system_prompt
        
        try:
            response = requests.post(
                url,
                headers=self.headers,
                files=files,
                data=data,
                timeout=60  # 60 second timeout
            )
            
            # Close the file
            files['file'][1].close()
            
            # Handle different response status codes
            if response.status_code == 200:
                return response.json()
            elif response.status_code == 400:
                error_detail = response.json().get('detail', 'Bad Request')
                raise requests.RequestException(f"Bad Request: {error_detail}")
            elif response.status_code == 413:
                raise requests.RequestException("File too large (max 50MB)")
            elif response.status_code == 429:
                error_detail = response.json().get('detail', {})
                if isinstance(error_detail, dict):
                    message = error_detail.get('message', 'Usage limit exceeded')
                    raise requests.RequestException(f"Usage limit exceeded: {message}")
                else:
                    raise requests.RequestException("Usage limit exceeded")
            elif response.status_code == 500:
                error_detail = response.json().get('detail', 'Internal Server Error')
                raise requests.RequestException(f"Internal Server Error: {error_detail}")
            else:
                response.raise_for_status()
                
        except requests.exceptions.Timeout:
            raise requests.RequestException("Request timed out")
        except requests.exceptions.ConnectionError:
            raise requests.RequestException("Connection error")
        except json.JSONDecodeError:
            raise requests.RequestException("Invalid JSON response")
        finally:
            # Ensure file is closed even if an exception occurs
            if 'files' in locals() and files['file'][1]:
                files['file'][1].close()


def main():
    """Example usage of the Meditrail OCR Client"""
    
    # Configuration
    API_KEY = "mt_LaRysGk4Ap9ytP29VKtF9E_8MECKcLcZSgRYLAF2KP5qr_UVPdCUqchtJ9fM7joiXX8z"
    
    # Initialize client
    client = MeditrailOCRClient(API_KEY)
    
    # Example 1: Process a medical image
    print("=== Example 1: Processing Medical Image ===")
    try:
        # Replace with your actual file path
        file_path = "sample_files/chest_xray.jpg"
        
        if os.path.exists(file_path):
            result = client.process_document(
                file_path=file_path,
                text="Chest X-ray examination",
                system_prompt="Extract key clinical findings and abnormalities"
            )
            
            print(f"Success! Document ID: {result.get('id')}")
            print(f"Clinical Relevance: {result.get('clinical_relevance')}")
            print(f"Doctor Names: {result.get('doctor_names')}")
            
            # Parse the response field (it's a JSON string)
            response_data = json.loads(result.get('response', '{}'))
            print(f"Document Type: {response_data.get('document_type', 'N/A')}")
            print(f"Summary: {response_data.get('summary', 'N/A')}")
            
            # Print metadata
            metadata = result.get('metadata', {})
            print(f"Original File: {metadata.get('original_file_name')}")
            print(f"File Size: {metadata.get('file_size')}")
            print(f"Page Count: {metadata.get('page_count')}")
            
        else:
            print(f"Sample file not found: {file_path}")
            print("Please add a sample medical document to test with.")
            
    except Exception as e:
        print(f"Error processing document: {e}")
    
    print("\n" + "="*50 + "\n")
    
    # Example 2: Process a PDF document
    print("=== Example 2: Processing PDF Document ===")
    try:
        # Replace with your actual file path
        file_path = "sample_files/prescription.pdf"
        
        if os.path.exists(file_path):
            result = client.process_document(
                file_path=file_path,
                text="Prescription document",
                system_prompt="Extract medication names, dosages, and doctor information"
            )
            
            print(f"Success! Document ID: {result.get('id')}")
            print(f"Clinical Relevance: {result.get('clinical_relevance')}")
            
            # Parse the response field
            response_data = json.loads(result.get('response', '{}'))
            print(f"Document Type: {response_data.get('document_type', 'N/A')}")
            print(f"Summary: {response_data.get('summary', 'N/A')}")
            
        else:
            print(f"Sample file not found: {file_path}")
            print("Please add a sample PDF document to test with.")
            
    except Exception as e:
        print(f"Error processing document: {e}")


if __name__ == "__main__":
    main()
